File: js/ai/src/session.ts
/**
 * Copyright 2026 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { getAsyncContext, z, type ActionContext } from '@genkit-ai/core';
import { EventEmitter } from '@genkit-ai/core/async';
import type { Registry } from '@genkit-ai/core/registry';
import * as fs from 'fs';
import * as fsp from 'fs/promises';
import * as path from 'path';
import { MessageData, MessageSchema } from './model-types.js';

import { PartSchema } from './model-types.js';

/**
 * Schema for tracking persistent artifacts generated during a session turn.
 */
export const ArtifactSchema = z.object({
  name: z.string().optional(),
  parts: z.array(PartSchema),
  metadata: z.record(z.any()).optional(),
});

/**
 * Artifact generated during a session turn.
 */
export type Artifact = z.infer<typeof ArtifactSchema>;

/**
 * Events signifying a session snapshot persistence point.
 */
export const SnapshotEventSchema = z.enum(['turnEnd', 'invocationEnd']);

/**
 * Event signifying a session snapshot persistence point.
 */
export type SnapshotEvent = z.infer<typeof SnapshotEventSchema>;

/**
 * Schema for session execution state.
 */
export const SessionStateSchema = z.object({
  messages: z.array(MessageSchema).optional(),
  custom: z.any().optional(),
  artifacts: z.array(ArtifactSchema).optional(),
  inputVariables: z.any().optional(),
});

/**
 * State persisted for a session across turns.
 */
export interface SessionState<S = unknown, I = unknown> {
  messages?: MessageData[];
  custom?: S;
  artifacts?: Artifact[];
  inputVariables?: I;
}

/**
 * The execution context provided to a snapshot callback.
 */
export interface SnapshotContext<S = unknown, I = unknown> {
  state: SessionState<S, I>;
  prevState?: SessionState<S, I>;
  turnIndex: number;
  event: 'turnEnd' | 'invocationEnd';
}

/**
 * Callback triggered before a snapshot is saved. Return false to reject persistence.
 */
export type SnapshotCallback<S = unknown, I = unknown> = (
  ctx: SnapshotContext<S, I>
) => boolean;

/**
 * Saved snapshot of a session's state at a given event point.
 */
export interface SessionSnapshot<S = unknown, I = unknown> {
  snapshotId: string;
  parentId?: string;
  createdAt: string;
  event: 'turnEnd' | 'invocationEnd';
  state: SessionState<S, I>;
  status?: 'pending' | 'done' | 'failed' | 'aborted';

  error?: {
    status: string;
    message: string;
    details?: any;
  };
}

/**
 * Options provided to the session store methods.
 */
export interface SessionStoreOptions {
  context?: ActionContext;
}

/**
 * Interface for persistent session snapshot storage.
 */
export interface SessionStore<S = unknown, I = unknown> {
  getSnapshot(
    snapshotId: string,
    options?: SessionStoreOptions
  ): Promise<SessionSnapshot<S, I> | undefined>;
  saveSnapshot(
    snapshot: SessionSnapshot<S, I>,
    options?: SessionStoreOptions
  ): Promise<void>;
  onSnapshotStateChange?(
    snapshotId: string,
    callback: (snapshot: SessionSnapshot<S, I>) => void,
    options?: SessionStoreOptions
  ): void | (() => void);
}

/**
 * State manager for a session turn, tracking messages, custom state, and artifacts.
 */
export class Session<S = unknown, I = unknown> extends EventEmitter {
  private state: SessionState<S, I>;
  private version: number = 0;

  constructor(initialState: SessionState<S, I>) {
    super();
    this.state = initialState;
  }

  /**
   * Returns a deep copy of the current session state.
   */
  getState(): SessionState<S, I> {
    return structuredClone(this.state);
  }

  /**
   * Retrieves all messages associated with the session.
   */
  getMessages(): MessageData[] {
    return this.state.messages || [];
  }

  /**
   * Appends a list of messages to the session.
   */
  addMessages(messages: MessageData[]) {
    this.state.messages = [...(this.state.messages || []), ...messages];
    this.version++;
  }

  /**
   * Overwrites the session messages.
   */
  setMessages(messages: MessageData[]) {
    this.state.messages = messages;
    this.version++;
  }

  /**
   * Retrieves the custom state of the session.
   */
  getCustom(): S | undefined {
    return this.state.custom;
  }

  /**
   * Updates the custom state of the session using a mutator function.
   */
  updateCustom(fn: (custom?: S) => S) {
    this.state.custom = fn(this.state.custom);
    this.version++;
  }

  /**
   * Retrieves the list of artifacts generated during the session.
   */
  getArtifacts(): Artifact[] {
    return this.state.artifacts || [];
  }

  /**
   * Adds artifacts to the session, deduplicating items by name.
   * Emits 'artifactAdded' for new artifacts and 'artifactUpdated' for replacements.
   */
  addArtifacts(artifacts: Artifact[]) {
    const existing = this.state.artifacts || [];
    const added: Artifact[] = [];
    const updated: Artifact[] = [];

    for (const a of artifacts) {
      if (a.name) {
        const idx = existing.findIndex((e) => e.name === a.name);
        if (idx >= 0) {
          existing[idx] = a;
          updated.push(a);
          continue;
        }
      }
      existing.push(a);
      added.push(a);
    }

    this.state.artifacts = existing;
    if (added.length + updated.length > 0) {
      this.version++;
    }
    for (const a of added) {
      this.emit('artifactAdded', a);
    }
    for (const a of updated) {
      this.emit('artifactUpdated', a);
    }
  }

  /**
   * Runs the provided function inside the session's context.
   */
  run<O>(fn: () => O) {
    return getAsyncContext().run('ai.session', this, fn);
  }

  /**
   * Gets the current mutation version of the session state.
   */
  getVersion(): number {
    return this.version;
  }
}

/**
 * In-memory implementation of persistent Session Store.
 */
export class InMemorySessionStore<S = unknown, I = unknown>
  implements SessionStore<S, I>
{
  private snapshots = new Map<string, SessionSnapshot<S, I>>();
  private listeners = new Map<
    string,
    Array<(snapshot: SessionSnapshot<S, I>) => void>
  >();

  async getSnapshot(
    snapshotId: string,
    options?: SessionStoreOptions
  ): Promise<SessionSnapshot<S, I> | undefined> {
    const snap = this.snapshots.get(snapshotId);
    if (!snap) return undefined;
    return structuredClone(snap);
  }

  async saveSnapshot(
    snapshot: SessionSnapshot<S, I>,
    options?: SessionStoreOptions
  ): Promise<void> {
    this.snapshots.set(snapshot.snapshotId, structuredClone(snapshot));
    const snapshotListeners = this.listeners.get(snapshot.snapshotId);
    if (snapshotListeners) {
      for (const listener of snapshotListeners) {
        listener(structuredClone(snapshot));
      }
    }
  }

  onSnapshotStateChange(
    snapshotId: string,
    callback: (snapshot: SessionSnapshot<S, I>) => void,
    options?: SessionStoreOptions
  ): void | (() => void) {
    if (!this.listeners.has(snapshotId)) {
      this.listeners.set(snapshotId, []);
    }
    this.listeners.get(snapshotId)!.push(callback);
    return () => {
      const list = this.listeners.get(snapshotId);
      if (list) {
        const index = list.indexOf(callback);
        if (index >= 0) list.splice(index, 1);
      }
    };
  }
}

/**
 * Utility to execute a function bound to a Session instance context.
 */
export function runWithSession<S = any, O = any>(
  registry: Registry,
  session: Session<S>,
  fn: () => O
): O {
  return getAsyncContext().run('ai.session', session, fn);
}

/**
 * Returns the Session instance active in the current context.
 */
export function getCurrentSession<S = any>(
  registry: Registry
): Session<S> | undefined {
  return getAsyncContext().getStore('ai.session');
}

/**
 * Error thrown during session execution.
 */
export class SessionError extends Error {
  constructor(msg: string) {
    super(msg);
  }
}

// Only UUID-shaped IDs are accepted to prevent path traversal.
const SAFE_ID_PATTERN =
  /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;

/**
 * A Node.js file-system backed session snapshot store.
 *
 * Each snapshot is persisted as a JSON file under `dirPath/<prefix>/<snapshotId>.json`.
 * Only UUID-formatted snapshot IDs are accepted to prevent path traversal.
 */
export class FileSessionStore<S = unknown, I = unknown>
  implements SessionStore<S, I>
{
  private dirPath: string;
  private maxPersistedChainLength?: number;
  private snapshotPathPrefix?: (
    snapshotId: string,
    options?: SessionStoreOptions
  ) => string;

  /**
   * @param dirPath Directory where snapshot JSON files are stored.
   * @param options.maxPersistedChainLength When set, snapshots older than this
   *   many entries in a chain are automatically deleted on each save.
   * @param options.snapshotPathPrefix Returns a sub-directory prefix per
   *   snapshot, useful for multi-tenant isolation. Defaults to `"global"`.
   */
  constructor(
    dirPath: string,
    options?: {
      maxPersistedChainLength?: number;
      snapshotPathPrefix?: (
        snapshotId: string,
        options?: SessionStoreOptions
      ) => string;
    }
  ) {
    this.dirPath = path.resolve(dirPath);
    fs.mkdirSync(this.dirPath, { recursive: true });
    this.maxPersistedChainLength = options?.maxPersistedChainLength;
    this.snapshotPathPrefix = options?.snapshotPathPrefix;
  }

  private validateId(snapshotId: string): void {
    if (!SAFE_ID_PATTERN.test(snapshotId)) {
      throw new Error(`Invalid snapshotId: "${snapshotId}"`);
    }
  }

  private async ensureDir(dir: string): Promise<void> {
    await fsp.mkdir(dir, { recursive: true });
  }

  private async getFilePath(
    snapshotId: string,
    options?: SessionStoreOptions
  ): Promise<string> {
    this.validateId(snapshotId);
    const prefix = this.snapshotPathPrefix
      ? this.snapshotPathPrefix(snapshotId, options)
      : 'global';
    const dir = path.join(this.dirPath, prefix);
    await this.ensureDir(dir);
    return path.join(dir, `${snapshotId}.json`);
  }

  async getSnapshot(
    snapshotId: string,
    options?: SessionStoreOptions
  ): Promise<SessionSnapshot<S, I> | undefined> {
    const filePath = await this.getFilePath(snapshotId, options);
    try {
      const fileContents = await fsp.readFile(filePath, 'utf-8');
      return JSON.parse(fileContents) as SessionSnapshot<S, I>;
    } catch (e: unknown) {
      if ((e as NodeJS.ErrnoException).code === 'ENOENT') return undefined;
      throw e;
    }
  }

  async saveSnapshot(
    snapshot: SessionSnapshot<S, I>,
    options?: SessionStoreOptions
  ): Promise<void> {
    const filePath = await this.getFilePath(snapshot.snapshotId, options);
    await fsp.writeFile(filePath, JSON.stringify(snapshot, null, 2), 'utf-8');

    if (this.maxPersistedChainLength && this.maxPersistedChainLength > 0) {
      let current: SessionSnapshot<S, I> | undefined = snapshot;
      const chain: string[] = [];

      while (current) {
        chain.push(current.snapshotId);
        if (current.parentId) {
          current = await this.getSnapshot(current.parentId, options);
        } else {
          break;
        }
      }

      if (chain.length > this.maxPersistedChainLength) {
        for (let i = this.maxPersistedChainLength; i < chain.length; i++) {
          const pathToDelete = await this.getFilePath(chain[i], options);
          await fsp.unlink(pathToDelete).catch((e: unknown) => {
            if ((e as NodeJS.ErrnoException).code !== 'ENOENT') throw e;
          });
        }
      }
    }
  }
}

File: go/ai/session.go
// Copyright 2024 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package ai

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"sync"

	"github.com/firebase/genkit/go/core"
	"github.com/firebase/genkit/go/internal/base"
	"github.com/google/uuid"
)

// Sessions are by default stored in memory only.
// Implement the Store interface to persist state.
type SessionStore interface {
	Get(sessionId string) (data SessionData, err error)
	Save(sessionId string, data SessionData) error
}

type SessionData struct {
	State       map[string]any        `json:"state,omitempty"`       // Any state that should be stored
	StateSchema map[string]any        `json:"stateschema,omitempty"` // Schema for state variables
	Threads     map[string][]*Message `json:"threads,omitempty"`     // Messages by thread name
}

type Session struct {
	id    string       // The session id
	data  *SessionData // The data for the session
	store SessionStore // The store for the session, defaults to in-memory storage
}

var sessionKey = base.NewContextKey[*Session]() // A session key

// NewSession creates a new session with the provided options.
// If no store is provided, it defaults to in-memory storage.
func NewSession(ctx context.Context, opts ...SessionOption) (session *Session, err error) {
	s := &Session{}

	sessOpts := &sessionOptions{}
	for _, opt := range opts {
		if err := opt.applySession(sessOpts); err != nil {
			return nil, core.NewError(core.INVALID_ARGUMENT, "ai.NewSession: error applying options: %v", err)
		}
	}

	s.store = &InMemorySessionStore{}
	if sessOpts.Store != nil {
		s.store = sessOpts.Store
	}

	s.id = uuid.New().String()
	if sessOpts.ID != "" {
		s.id = sessOpts.ID
	}

	if sessOpts.Data != nil {
		s.data = sessOpts.Data
	}

	if s.data == nil {
		s.data = &SessionData{
			Threads: make(map[string][]*Message),
		}
	}

	// Set default state
	if sessOpts.Schema != nil {
		s.data.StateSchema = sessOpts.Schema
		state, err := getState(s.data.StateSchema, sessOpts.DefaultState)
		if err != nil {
			return nil, err
		}

		s.data.State = state
	}

	// Initialize session in store
	s.store.Save(s.id, *s.data)

	return s, nil
}

// Load loads sessiondata from store, and returns a session.
func LoadSession(ctx context.Context, sessionId string, store SessionStore) (session *Session, err error) {
	sessionData, err := store.Get(sessionId)
	if err != nil {
		return nil, err
	}

	session = &Session{
		id:    sessionId,
		data:  &sessionData,
		store: store,
	}

	return session, nil
}

// GetID returns the session id.
func (s *Session) GetID() string {
	return s.id
}

// GetData returns the sessions data
func (s *Session) GetData() (SessionData, error) {
	return s.store.Get(s.id)
}

// UpdateState takes any data and sets it as state in the store.
func (s *Session) UpdateState(state any) error {
	var err error

	// Ensure sessionData is up to date
	sessionData, err := s.store.Get(s.id)
	if err != nil {
		return err
	}
	s.data = &sessionData

	// Use schema to get state
	if s.data.StateSchema != nil {
		s.data.State, err = getState(s.data.StateSchema, state)
		if err != nil {
			return err
		}
	} else {
		data, err := json.Marshal(state)
		if err != nil {
			return err
		}

		err = json.Unmarshal(data, &s.data.State)
		if err != nil {
			return fmt.Errorf("genkit.UpdateState: %w, state doesn't match existing schema", err)
		}
	}

	return s.store.Save(s.id, *s.data)
}

// UpdateMessages takes a threadName and a slice of messages.
func (s *Session) UpdateMessages(threadName string, messages []*Message) error {
	// Ensure sessionData is up to date
	sessionData, err := s.store.Get(s.id)
	if err != nil {
		return err
	}
	s.data = &sessionData

	s.data.Threads[threadName] = messages
	return s.store.Save(s.id, *s.data)
}

// Set current session in context.
func (s *Session) SetContext(ctx context.Context) context.Context {
	return sessionKey.NewContext(ctx, s)
}

// Find current session in context, returns session if found.
func SessionFromContext(ctx context.Context) (session *Session, err error) {
	if s := sessionKey.FromContext(ctx); s != nil {
		return s, nil
	}

	return nil, errors.New("genkit.SessionFromContext: session not found")
}

// Helper function to derive state from schema.
func getState(schema map[string]any, state any) (map[string]any, error) {
	data, err := json.Marshal(state)
	if err != nil {
		return nil, err
	}

	err = json.Unmarshal(data, &schema)
	if err != nil {
		return nil, fmt.Errorf("genkit.getState: %w, only structs are allowed as types", err)
	}

	return schema, nil
}

// Default in-memory session store.
type InMemorySessionStore struct {
	data map[string]SessionData // The session data keyed by session id
	mu   sync.RWMutex
}

func (s *InMemorySessionStore) Get(sessionId string) (data SessionData, err error) {
	s.mu.RLock()
	defer s.mu.RUnlock()

	if _, ok := s.data[sessionId]; !ok {
		return data, errors.New("genkit.InMemorySessionStore.Get: session not found")
	}
	return s.data[sessionId], nil
}

func (s *InMemorySessionStore) Save(sessionId string, data SessionData) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	if _, ok := s.data[sessionId]; !ok {
		s.data = make(map[string]SessionData)
	}
	s.data[sessionId] = data
	return nil
}
